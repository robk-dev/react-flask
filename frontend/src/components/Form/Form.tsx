import React, { FC } from 'react';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import { makeStyles } from '@material-ui/core/styles';

import { useForm } from 'react-hook-form';
import { ErrorMessage } from '@hookform/error-message';

import axios from '../../axios';
import { Data } from '../Table/Table';

const useStyles = makeStyles((theme) => ({
    paper: {
        marginTop: theme.spacing(2),
    },
    form: {
        width: '100%',
        marginTop: theme.spacing(1),
    },
    textField: {
        marginLeft: theme.spacing(1),
        marginRight: theme.spacing(1),
        width: 200,
    },
    submit: {
        margin: theme.spacing(3, 0, 2),
        backgroundColor: '#00AF66'
    }
}));

type FormInputs = {
    name: string;
    email: string;
    date: Date;
    kids: number;
};

interface FormProps {
    setData(data: Data[]): void
}

export const Form: FC<FormProps> = ({ setData }) => {
    const classes = useStyles();


    const onSubmit = async (form: FormInputs) => {
        try {
            const { data } = await axios.post('/api/users', form);
            setData(data);
        } catch (error) {
            console.error(error);
        }
    };

    let today = new Date();
    const max_18_years_ago = new Date(today.setFullYear(today.getFullYear() - 18)).toISOString().split('T')[0];

    const { register, handleSubmit, errors } = useForm<FormInputs>({
        criteriaMode: "all"
    });

    return (
        <form className={classes.form} onSubmit={handleSubmit(onSubmit)}>
            <TextField
                variant="outlined"
                margin="normal"
                required
                fullWidth
                id="name"
                label="Name"
                name="name"
                autoComplete="name"
                autoFocus
                inputRef={register({
                    required: "NAME IS REQUIRED",
                    minLength: { value: 1, message: 'PEOPLE\'S NAMES CONTAIN AT LEAST ONE CHARACTER' },
                    maxLength: { value: 100, message: '>100 CHARS FOR NAME IS PROBABLY TOO MUCH' }
                })}
            />
            <ErrorMessage errors={errors} name="name" />
            <TextField
                variant="outlined"
                margin="normal"
                fullWidth
                required
                id="email"
                label="Email"
                name="email"
                autoComplete="email"
                inputRef={register({
                    required: "EMAIL REQUIRED",
                    pattern: {
                        value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
                        message: "INVALID EMAIL ADDRESS"
                    },
                    maxLength: { value: 100, message: '>100 CHARS FOR EMAIL IS PROBABLY TOO MUCH' }
                })}
            />
            <ErrorMessage errors={errors} name="email" />
            <TextField
                variant="outlined"
                margin="normal"
                fullWidth
                required
                id="kids"
                type="number"
                label="Number of children"
                inputProps={{
                    max: 9999,
                    min: 0
                }}
                name="kids"
                inputRef={register({
                    required: "NUMBER OF CHILDREN IS REQUIRED"
                })} />
            <ErrorMessage errors={errors} name="kids" />
            <TextField
                id="date"
                label="Birthday"
                type="date"
                name='date'
                required
                fullWidth
                InputProps={{
                    inputProps: { max: max_18_years_ago, min: '1900-01-01' }
                }}
                defaultValue={max_18_years_ago}
                inputRef={register({
                    required: "BIRTHDAY IS REQUIRED",
                })}
            />
            <ErrorMessage errors={errors} name="date" />


            <Button
                type="submit"
                fullWidth
                variant="contained"
                color="primary"
                className={classes.submit}
            >
                Submit
                    </Button>
        </form >
    );
}