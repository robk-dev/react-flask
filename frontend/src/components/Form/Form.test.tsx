import React from 'react';

import { act, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Form } from './Form';

describe('Form Test Suite', () => {
    let container: any = null;

    const setData = () => {};
    beforeEach(() => {
        container = document.createElement("div");
        document.body.appendChild(container);
    });
    afterEach(() => {
        container.remove();
        container = null;
    });

    test('Form should render without crashing', async () => {
        await act(async () => {
            render(<Form setData={setData} />, container);
        });
    });
    test('Each input should have a label', async () => {
        await act(async () => {
            render(<Form setData={setData}/>, container);
        });
        const name_label = screen.getByLabelText(/Name/i);
        expect(name_label).toBeInTheDocument();
        const email_label = screen.getByLabelText(/Email/i);
        expect(email_label).toBeInTheDocument();
        const children_label = screen.getByLabelText(/Number of children/i);
        expect(children_label).toBeInTheDocument();
        const birthday_label = screen.getByLabelText(/Birthday/i);
        expect(birthday_label).toBeInTheDocument();
    });

    test('Form tree should match snapshot one', () => {
        const table = renderer.create(<Form setData={setData}/>)
        const tree = table.toJSON();
        expect(tree).toMatchSnapshot();
    });
})