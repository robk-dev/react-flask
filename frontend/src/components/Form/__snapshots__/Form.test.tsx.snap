// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Test Suite Form tree should match snapshot one 1`] = `
<form
  className="makeStyles-form-18"
  onSubmit={[Function]}
>
  <div
    className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
      data-shrink={false}
      htmlFor="name"
      id="name-label"
    >
      Name
      <span
        aria-hidden={true}
        className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoComplete="name"
        autoFocus={true}
        className="MuiInputBase-input MuiOutlinedInput-input"
        disabled={false}
        id="name"
        name="name"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={true}
        type="text"
      />
      <fieldset
        aria-hidden={true}
        className="PrivateNotchedOutline-root-21 MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="PrivateNotchedOutline-legendLabelled-23"
        >
          <span>
            Name
             *
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
      data-shrink={false}
      htmlFor="email"
      id="email-label"
    >
      Email
      <span
        aria-hidden={true}
        className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoComplete="email"
        autoFocus={false}
        className="MuiInputBase-input MuiOutlinedInput-input"
        disabled={false}
        id="email"
        name="email"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={true}
        type="text"
      />
      <fieldset
        aria-hidden={true}
        className="PrivateNotchedOutline-root-21 MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="PrivateNotchedOutline-legendLabelled-23"
        >
          <span>
            Email
             *
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
      data-shrink={false}
      htmlFor="kids"
      id="kids-label"
    >
      Number of children
      <span
        aria-hidden={true}
        className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiOutlinedInput-input"
        disabled={false}
        id="kids"
        max={9999}
        min={0}
        name="kids"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={true}
        type="number"
      />
      <fieldset
        aria-hidden={true}
        className="PrivateNotchedOutline-root-21 MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="PrivateNotchedOutline-legendLabelled-23"
        >
          <span>
            Number of children
             *
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled Mui-required Mui-required"
      data-shrink={true}
      htmlFor="date"
      id="date-label"
    >
      Birthday
      <span
        aria-hidden={true}
        className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
      >
         
        *
      </span>
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input"
        defaultValue="2002-11-12"
        disabled={false}
        id="date"
        max="2002-11-12"
        min="1900-01-01"
        name="date"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={true}
        type="date"
      />
    </div>
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submit-20 MuiButton-containedPrimary MuiButton-fullWidth"
    disabled={false}
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="submit"
  >
    <span
      className="MuiButton-label"
    >
      Submit
    </span>
  </button>
</form>
`;
